### PR TITLE
Added errors{'raise','ignore'} for keys not found in meta for json_normalize

### PR DESCRIPTION
**Pull request by dickreuter**
Continuation of https://github.com/pandas-dev/pandas/pull/14505

Added errors{'raise','ignore'} for keys not found in meta for json_normalize

### CODE REVIEW

**Comment by dickreuter** [comment]
Cleaning it up, should be better now. Is there a new whatsnew file so I can add the comments? What's the next version number?

**Comment by dickreuter** [comment]
Why is nobody merging this?

File: pandas/io/json.py [hunk 1]
```diff
@@ -740,6 +742,8 @@ def json_normalize(data, record_path=None, meta=None,
         If True, prefix records with dotted (?) path, e.g. foo.bar.field if
         path to records is ['foo', 'bar']
     meta_prefix : string, default None
+    error: {'raise', 'ignore'}, default 'raise'
```
Comment (jreback) [inline pandas/io/json.py]: need a version in 0.19.2 tag

File: doc/source/whatsnew/v0.20.0.txt [hunk 2]
```diff
@@ -80,3 +80,4 @@ Performance Improvements
 
 Bug Fixes
 ~~~~~~~~~
+- Enhancement in ``pandas.io.json.json_normalize``If meta keys are not always present a new option to set errors='ignore' has been implemented (:issue:`14505`)
```
Comment (jreback) [inline doc/source/whatsnew/v0.20.0.txt]: you can put in enhancements section

Comment (jreback) [inline doc/source/whatsnew/v0.20.0.txt]: space after quotes and before If. put ``errors='ignore'`` in double-backticks.

File: pandas/io/json.py [hunk 3]
```diff
@@ -839,7 +843,14 @@ def _recursive_extract(data, path, seen_meta, level=0):
...
                     else:
-                        meta_val = _pull_field(obj, val[level:])
+                        try:
+                            meta_val = _pull_field(obj, val[level:])
+                        except KeyError as e:
+                            if errors == 'ignore':
+                                meta_val = np.nan
+                            else:
+                                raise KeyError(
+                                    "Try running with errors='ignore' as the following key may not always be present: " + str(e))
```
Comment (jreback) [inline pandas/io/json.py]: this won't pass linting

File: pandas/io/tests/json/test_json_norm.py [hunk 4]
```diff
@@ -225,6 +225,59 @@ def test_nested_flattens(self):
 
         self.assertEqual(result, expected)
 
+
+    def test_json_normalise_fix(self):
+        # issue 14505
```
Comment (jreback) [inline pandas/io/tests/json/test_json_norm.py]: put a comment explaining what this is. rename test -> test_json_normalize_errors

**Comment by jreback** [comment]
@dickreuter we have 107 open PR's. very few people do code review. simply ping if you want someone to look.

**Comment by dickreuter** [comment]
Ok great. Please have a look. I think this can now be merged. 






On Thu, Nov 17, 2016 at 11:29 PM +0000, "Joris Van den Bossche" <notifications@github.com> wrote:












@dickreuter just so you know: we don't get notifications when you push new code, only from comments. So it's always good to put a comment after you pushed new code to indicated that you updated. And as @jreback says, pinging like you now did always helps

**Comment by jorisvandenbossche** [comment]
@dickreuter Did you push the changes? As there is still a merge conflict and there is no new commit since @jreback 's comments

File: pandas/io/json.py (hunk 3 above)
Comment (dickreuter) [inline pandas/io/json.py]: Why not? What do I need to change?

Comment (jorisvandenbossche) [inline pandas/io/json.py]: We check for PEP8, so this line will be too long. Normally you can set your editor to check for those things, but can also see the output of the check on travis (3d build): https://travis-ci.org/pandas-dev/pandas/jobs/174180562 (at the bottom)

```
pandas/io/json.py:746:80: E501 line too long (85 > 79 characters)
pandas/io/json.py:853:80: E501 line too long (129 > 79 characters)
pandas/io/tests/json/test_json_norm.py:229:5: E303 too many blank lines (2)
pandas/io/tests/json/test_json_norm.py:271:80: E501 line too long (81 > 79 characters)
pandas/io/tests/json/test_json_norm.py:272:80: E501 line too long (104 > 79 characters)
pandas/io/tests/json/test_json_norm.py:273:17: E225 missing whitespace around operator
pandas/io/tests/json/test_json_norm.py:274:10: E128 continuation line under-indented for visual indent
pandas/io/tests/json/test_json_norm.py:275:10: E128 continuation line under-indented for visual indent
pandas/io/tests/json/test_json_norm.py:276:10: E128 continuation line under-indented for visual indent
pandas/io/tests/json/test_json_norm.py:277:10: E128 continuation line under-indented for visual indent
```

**Comment by dickreuter** [comment]
Just pushed the changes as instructed.

**Comment by dickreuter** [comment]
Anything else I need to fix in this or can this be merged?

**Comment by dickreuter** [comment]
Just did another rebase and pushed again. Can this be merged now? thanks

File: doc/source/whatsnew/v0.20.0.txt [hunk 5]
```diff
@@ -49,6 +49,7 @@ Other enhancements
 ^^^^^^^^^^^^^^^^^^
 
 - ``pd.read_excel`` now preserves sheet order when using ``sheetname=None`` (:issue:`9930`)
+- ``pandas.io.json.json_normalize`` If meta keys are not always present a new option to set errors="ignore" (:issue:`14583`)
```
Comment (jreback) [inline doc/source/whatsnew/v0.20.0.txt]: In ``pandas.io.json.json_normalize`` gained the option ``errors='ignore'|raise`` (in double-backticks); the default is ``raise`` and is backward compatible.

File: doc/source/whatsnew/v0.20.0.txt [hunk 6]
```diff
@@ -104,4 +105,4 @@ Performance Improvements
 .. _whatsnew_0200.bug_fixes:
 
 Bug Fixes
-~~~~~~~~~
```
Comment (jreback) [inline doc/source/whatsnew/v0.20.0.txt]: revert this

File: pandas/io/json.py (hunk 1 above)
Comment (jreback) [inline pandas/io/json.py]: need a version added tag

File: pandas/io/json.py [hunk 7]
```diff
@@ -740,6 +742,8 @@ def json_normalize(data, record_path=None, meta=None,
         If True, prefix records with dotted (?) path, e.g. foo.bar.field if
         path to records is ['foo', 'bar']
     meta_prefix : string, default None
+    error: {'raise', 'ignore'}, default 'raise'
+        * ignore: will ignore keyErrors if keys listed in meta are not always present
```
Comment (jreback) [inline pandas/io/json.py]: add an entry for raise

Comment (jreback) [inline pandas/io/json.py]: will ignore ``KeyError``, if the keys passed in meta are not all present

File: pandas/io/tests/json/test_json_norm.py [hunk 8]
```diff
@@ -225,6 +225,65 @@ def test_nested_flattens(self):
 
         self.assertEqual(result, expected)
 
+
+    def test_json_normalize_errors(self):
+        # If meta keys are not always present a new option to set errors='ignore' has been implemented (:issue:`14583`)
```
Comment (jreback) [inline pandas/io/tests/json/test_json_norm.py]: GH14583

**Comment by jreback** [comment]
just some doc comments, otherwise lgtm.

**Comment by dickreuter** [comment]
Implemented doc amendments as suggested. Let me know if anything else needs to be changed. thanks

File: pandas/io/json.py [hunk 9]
```diff
@@ -740,6 +742,9 @@ def json_normalize(data, record_path=None, meta=None,
         If True, prefix records with dotted (?) path, e.g. foo.bar.field if
         path to records is ['foo', 'bar']
     meta_prefix : string, default None
+    error: {'raise', 'ignore'}, default 'raise'
+        * ignore: will ignore KeyError if keys listed in meta are not always present
+        * raise: will raise KeyError if keys listed in meta are not always present
```
Comment (jorisvandenbossche) [inline pandas/io/json.py]: I suspect those two lines above are too long (> 80, we check for PEP8, which will let travis fail)

File: pandas/io/json.py [hunk 10]
```diff
@@ -839,7 +844,13 @@ def _recursive_extract(data, path, seen_meta, level=0):
...
                         meta_val = seen_meta[key]
                     else:
-                        meta_val = _pull_field(obj, val[level:])
+                        try:
+                            meta_val = _pull_field(obj, val[level:])
+                        except KeyError as e:
+                            if errors == 'ignore':
+                                meta_val = np.nan
+                            else:
+                                raise KeyError("Try running with errors='ignore' as key %s is not always present.", e)
```
Comment (jorisvandenbossche) [inline pandas/io/json.py]: same here, for PEP8 line too long

File: pandas/io/tests/json/test_json_norm.py [hunk 11]
```diff
@@ -225,6 +225,65 @@ def test_nested_flattens(self):
 
         self.assertEqual(result, expected)
 
+
+    def test_json_normalize_errors(self):
+        # GH14583: If meta keys are not always present a new option to set errors='ignore' has been implemented
```
Comment (jorisvandenbossche) [inline pandas/io/tests/json/test_json_norm.py]: here and in other places below line length is also a problem

**Review by jorisvandenbossche — COMMENTED** [review]
Some PEP8 style issues, otherwise looks good!

**Comment by jorisvandenbossche** [comment]
@dickreuter You still have some style issues (that's the reason that travis is failing). You can always look in the log of the third travis build to see the issues (https://travis-ci.org/pandas-dev/pandas/jobs/182143127, scroll to the bottom):

```
Linting  *.py
pandas/io/json.py:750:1: W293 blank line contains whitespace
pandas/io/json.py:857:80: E501 line too long (81 > 79 characters)
pandas/io/json.py:858:80: E501 line too long (85 > 79 characters)
pandas/io/tests/json/test_json_norm.py:229:5: E303 too many blank lines (2)
pandas/io/tests/json/test_json_norm.py:266:80: E501 line too long (81 > 79 characters)
pandas/io/tests/json/test_json_norm.py:267:80: E501 line too long (87 > 79 characters)
pandas/io/tests/json/test_json_norm.py:269:17: E225 missing whitespace around operator
pandas/io/tests/json/test_json_norm.py:270:10: E128 continuation line under-indented for visual indent
pandas/io/tests/json/test_json_norm.py:271:10: E128 continuation line under-indented for visual indent
pandas/io/tests/json/test_json_norm.py:272:10: E128 continuation line under-indented for visual indent
pandas/io/tests/json/test_json_norm.py:273:10: E128 continuation line under-indented for visual indent
pandas/io/tests/json/test_json_norm.py:278:17: E128 continuation line under-indented for visual indent
pandas/io/tests/json/test_json_norm.py:278:80: E501 line too long (86 > 79 characters)
pandas/io/tests/json/test_json_norm.py:279:28: E127 continuation line over-indented for visual indent
pandas/io/tests/json/test_json_norm.py:279:80: E501 line too long (87 > 79 characters)
Linting *.py DONE
```

(it also a good idea to set up your IDE to warn for this)

**Comment by dickreuter** [comment]
I believe this is fixed now. Some checks are still not successful but I can't find any reason why. Any suggestions why? All it says is:

The command "ci/lint.sh" exited with 1.
0.00s$ echo "script done"
script done

**Comment by jreback** [comment]
git diff master | flake8 -diff

**Comment by jorisvandenbossche** [comment]
@dickreuter You can also always look at the third travis build (https://travis-ci.org/pandas-dev/pandas/jobs/183014087, scroll to bottom), which says "pandas/io/json.py:857:45: E211 whitespace before '('"

**Comment by dickreuter** [comment]
Linting still seems to fail, even though when I do "git diff master | flake8 --diff" I only get problems in v0.20.0.txt, all the rest is fine. Any suggestions how I can find the remaining issues?

**Comment by jorisvandenbossche** [comment]
@dickreuter It was not linting this time, the mac build failed.

**Comment by dickreuter** [comment]
And it is related to my code? I can't find any details.

**Comment by jreback** [comment]
@dickreuter you need to rebase on master

``git rebase -i origin/master`` then
``git push -f``

**Comment by dickreuter** [comment]
Now all checked have passed.

**Comment by jreback** [comment]
thanks!

**Pull request closed without merging**